src/Passes/Source/Profiles/BaseProfile.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Passes/ExpandStaticAllocation/ExpandStaticAllocation.hpp"
+#include "Passes/QirAllocationAnalysis/QirAllocationAnalysis.hpp"
+#include "Passes/TransformationRule/TransformationRule.hpp"
+#include "Profiles/BaseProfile.hpp"
+#include "Rules/Factory.hpp"
+
+#include "Llvm/Llvm.hpp"
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    llvm::ModulePassManager BaseProfile::createGenerationModulePass(
+        llvm::PassBuilder&                    pass_builder,
+        llvm::PassBuilder::OptimizationLevel& optimisation_level,
+        bool                                  debug)
+    {
+        auto ret = pass_builder.buildPerModuleDefaultPipeline(optimisation_level);
+        // buildPerModuleDefaultPipeline buildModuleOptimizationPipeline
+        auto function_pass_manager = pass_builder.buildFunctionSimplificationPipeline(
+            optimisation_level, llvm::PassBuilder::ThinLTOPhase::None, debug);
+
+        auto inliner_pass =
+            pass_builder.buildInlinerPipeline(optimisation_level, llvm::PassBuilder::ThinLTOPhase::None, debug);
+
+        // TODO(tfr): Maybe this should be done at a module level
+        function_pass_manager.addPass(ExpandStaticAllocationPass());
+
+        RuleSet rule_set;
+
+        // Defining the mapping
+        auto factory = RuleFactory(rule_set);
+
+        factory.useStaticQubitArrayAllocation();
+        factory.useStaticQubitAllocation();
+        factory.useStaticResultAllocation();
+
+        factory.optimiseBranchQuatumOne();
+        //  factory.optimiseBranchQuatumZero();
+
+        factory.disableReferenceCounting();
+        factory.disableAliasCounting();
+        factory.disableStringSupport();
+
+        function_pass_manager.addPass(TransformationRulePass(std::move(rule_set)));
+
+        // Eliminate dead code
+        function_pass_manager.addPass(llvm::DCEPass());
+        function_pass_manager.addPass(llvm::ADCEPass());
+
+        //  function_pass_manager.addPass(llvm::createCalledValuePropagationPass());
+        // function_pass_manager.addPass(createSIFoldOperandsPass());
+
+        // Legacy passes:
+        // https://llvm.org/doxygen/group__LLVMCTransformsIPO.html#ga2ebfe3e0c3cca3b457708b4784ba93ff
+
+        // https://llvm.org/docs/NewPassManager.html
+        // modulePassManager.addPass(createModuleToCGSCCPassAdaptor(...));
+        // InlinerPass()
+
+        // auto &cgpm = inliner_pass.getPM();
+        // cgpm.addPass(llvm::ADCEPass());
+
+        // CGPM.addPass(createCGSCCToFunctionPassAdaptor(createFunctionToLoopPassAdaptor(LoopFooPass())));
+        // CGPM.addPass(createCGSCCToFunctionPassAdaptor(FunctionFooPass()));
+
+        ret.addPass(createModuleToFunctionPassAdaptor(std::move(function_pass_manager)));
+
+        // TODO(tfr): Not available in 11
```

Comment:
The comment here is not very clear; what is the purpose of keeping the commented out line of code - should it be uncommented if we updated the llvm version? What should be removed instead?
I have similar questions for the other commented out code. If there is any commented out code left here, then it should be clear when it should be uncommented and if anything else should be removed instead.

Reply:
Yes, it means, this code does not work in LLVM 11.  I will try to clarify this a bit, but it relates to my question about whether we would soon upgrade to 12 as the way to get a CG pass manager is somewhat hacky in 11 whereas it looks nice and clean in 12.

Reply:
I've made it more clear.